Introduce GCLock for cachedNode

KimKyungup opened the pull request:
## Proposed changes

This PR resolves https://github.com/klaytn/klaytn/issues/432 without increasing `triesInMemory`.
This added a lock for GC cachedNode.

During `traceBlock`, this GCLock can prevent to gc the cachedNode in the stateDB.

Also this added some exception code for error case of `AsMessageWithAccountKeyPicker` during `traceBlock`, `traceXXXX`.

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- https://github.com/klaytn/klaytn/issues/432

## Further comments

After this PR, I will rename caches related with state trie later.

kjhman21 commented on storage/statedb/database.go:
```diff
@@ -339,6 +340,16 @@ func (db *Database) DiskDB() database.DBManager {
+}
+
+// UnLockGCCachedNode unlocks the GC lock of CachedNode.
+func (db *Database) UnLockGCCachedNode() {
```
We usually use 'Unlock' rather than 'UnLock'. Please check out the below call.

kjhman21 commented:
1. Can we make a test case for this? e.g., make a fault when statedb is garbage collected?
2. I'm not sure a better solution, but using a lock in this way can cause unintended errors when others change this implementation. Is there any better way of using gc lock..?

ehnuje commented on node/cn/api_tracer.go:
```diff
@@ -454,10 +460,23 @@ func (api *PrivateDebugAPI) traceBlock(ctx context.Context, block *types.Block,
+		}
+		logger.Debug("Get stateDB by computeStateDB", "block", block.NumberU64())
+	} else {
+		logger.Debug("Get stateDB from stateCache", "block", block.NumberU64(), "&stateDB", fmt.Sprintf("%p", statedb))
```
Is it helpful to leave the pointer value of stateDB?

KimKyungup replied:
I used this during debugging. Now I don't need. I will remove. Thanks.

ehnuje commented on blockchain/blockchain.go:
```diff
@@ -1045,23 +1059,51 @@ func (bc *BlockChain) writeStateTrie(block *types.Block, state *state.StateDB) e
+		// trigger to GC stateDB cache
+		select {
+		case bc.gcTrigger <- block.NumberU64():
+		default:
```
If block number is not able to be passed to the `bc.gcTrigger` channel, it does nothing. Is it what you expected?

KimKyungup replied:
Yes it is. This routine is main routine. 
So if this trigger can't pass the signal and it waits, the main routine has some problem. 

I don't want this situation. Therefore this signal is not buffered. 
But, if a trigger is ignored, it will try again in next block.

kjhman21 replied:
Hmm... but you need to take care of this because it may change original behavior. Please add some logs for the unwanted case.

ehnuje replied:
I also agree with the @kjhman21 's point that we need to leave a log here. 
Log would be needed for the both cases.

KimKyungup replied:
I added a log here and inside of the loop. Thanks.

kjhman21 replied:
What about adding a log for `default:` case? I think this is unusual and in that case one block number can be missing. Can we use `logger.Error()` for `default:` case?

KimKyungup replied:
That default situation is not error case. 
It is intended case.
So How about `logger.Warn()`?

ehnuje replied:
logger.Warn() seems to be enough for this case (unless it skips too much)

KimKyungup replied:
I added warn log. Thanks. :)

KimKyungup commented:
> 2\. I'm not sure a better solution, but using a lock in this way can cause unintended errors when others change this implementation. Is there any better way of using gc lock..?

KimKyungup commented:
> 1. Can we make a test case for this? e.g., make a fault when statedb is garbage collected?
- If a stateDB has a fault caused by any reason, getObject() will try to find the object from TrieNodeCahce (BigCache) or LevelDB. If it is impossible to get the object from them, it will return nil and it can make some same situation of the issue. Let's consider more test cases.


> 2. I'm not sure a better solution, but using a lock in this way can cause unintended errors when others change this implementation. Is there any better way of using gc lock..?

Before this solution, I thought some solution like below.
- TrieInMemory 4 -> 128 : It will increase memory and increase `Cap()` also. More `Cap()` can decrease performance.
- Every `traceBlock` or `traceSomething` always regenerate the state before the snapshot. 
: This makes some decrease the performance of traceBlock.

I tried some other solution but I found the key of this problem is GC.
This change doesn't make some effect for main routine.  Only 1 lock/unlock routine is added in Dereference(). There is no other routine related with the lock.

But this change make an effect when trace APIs are called.

ehnuje commented on blockchain/blockchain.go:
```diff
@@ -444,6 +447,17 @@ func (bc *BlockChain) StateAt(root common.Hash) (*state.StateDB, error) {
+		return state.New(root, bc.stateCache)
+	}
+	bc.UnlockGCCachedNode()
+	return nil, errors.New("the node does not exist in state cache")
```
Please define a variable for this error.

KimKyungup replied:
Reflected. :)

ehnuje commented on blockchain/blockchain.go:
```diff
@@ -444,6 +447,17 @@ func (bc *BlockChain) StateAt(root common.Hash) (*state.StateDB, error) {
 }
 
+// StateAtWithGCLock returns a new mutable state based on a particular point in time with read lock of the state nodes.
+func (bc *BlockChain) StateAtWithGCLock(root common.Hash) (*state.StateDB, error) {
```
I'm a bit worried about this function since the caller of the function should take care of unlocking.

KimKyungup replied:
Yes. You should pay attention to the use of this feature.
If you are too worried about this, we have the option of moving this lock / error handling code outward.

kjhman21 replied:
> Yes. You should pay attention to the use of this feature.

Actually, this is what I addressed before. Taking care of unlocking outside of this function is very error-prone.

ehnuje commented on blockchain/blockchain.go:
```diff
@@ -444,6 +447,17 @@ func (bc *BlockChain) StateAt(root common.Hash) (*state.StateDB, error) {
+	bc.LockGCCachedNode()
+	exist := bc.stateCache.TrieDB().DoesExistCachedNode(root)
+	if exist {
+		return state.New(root, bc.stateCache)
```
What if `state.New` returns an error? We don't unlock here, but the caller also doesn't unlock the lock.

KimKyungup replied:
Thanks. Reflected. :)

KimKyungup commented:
@kjhman21 @ehnuje @winnie-byun 
I added 6699fa9 for preventing data race of `triegc             *prque.Prque`.

As you know, I changed the gc routine as a goroutine.
This routine makes a data race with `bc.triegc.Push(root, -float32(block.NumberU64()))` in `writeStateTrie` and another GC logic in `blockchain.Stop()`.
So I moved the push logic into the loop routine and applied wait group to the loop.

But, I think it is a bit complex. PTAL.

ehnuje commented on blockchain/blockchain.go:
```diff
@@ -179,20 +188,21 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
+		cacheConfig:        cacheConfig,
+		db:                 db,
+		triegc:             prque.New(),
+		chPushBlockGCPrque: make(chan gcBlock, 1000),
```
How about using simpler name like below?
```suggestion
		gcBlockCh: make(chan gcBlock, 1000),
```

ehnuje commented on storage/statedb/database.go:
```diff
@@ -98,6 +98,7 @@ type Database struct {
 	gctime  time.Duration      // Time spent on garbage collection since last commit
 	gcnodes uint64             // Nodes garbage collected since last commit
 	gcsize  common.StorageSize // Data storage garbage collected since last commit
+	gcLock  sync.RWMutex       // Lock for preventing to garbage collect cachedNode without flushing.
```
Minor suggestion, but just for consistency.
```suggestion
	gcLock  sync.RWMutex       // Lock for preventing to garbage collect cachedNode without flushing
```

ehnuje commented on storage/statedb/database.go:
```diff
@@ -431,13 +442,15 @@ func (db *Database) node(hash common.Hash) node {
 	if node != nil {
 		return node.obj(hash)
 	}
+
```
Changes in L#437 and L#453 seems unnecessary.

KimKyungup replied:
I added lines for readability.

ehnuje commented on storage/statedb/database.go:
```diff
@@ -474,6 +487,13 @@ func (db *Database) Node(hash common.Hash) ([]byte, error) {
 	return enc, err
 }
 
+// DoesExistCachedNode return if the noce exist from cached trie node in memory.
```
```suggestion
// DoesExistCachedNode returns if the node exists on the cached trie node in memory.
```

ehnuje commented on storage/statedb/database.go:
```diff
@@ -474,6 +487,13 @@ func (db *Database) Node(hash common.Hash) ([]byte, error) {
 }
 
+// DoesExistCachedNode return if the noce exist from cached trie node in memory.
+func (db *Database) DoesExistCachedNode(hash common.Hash) bool {
```
Don't we need to look for the node in the trie node cache?
Also, it seems that you need to lock the `db.lock.RLock()` according to the line 475~477.

KimKyungup replied:
I added Lock. Thanks. :)

ehnuje commented on node/cn/handler_test.go:
```diff
@@ -112,6 +112,24 @@ func newBlock(blockNum int) *types.Block {
 	return block
 }
 
+func newBlockWithParentHash(blockNum int, parentHash common.Hash) *types.Block {
```
Why is this function needed?

KimKyungup replied:
In `TestPrivateDebugAPI_computeTxEnv` unit test, there is a logic to check the parent hash of the block.

ehnuje commented on node/cn/api_tracer.go:
```diff
@@ -780,16 +826,24 @@ func (api *PrivateDebugAPI) computeTxEnv(blockHash common.Hash, txIndex int, ree
-		return nil, vm.Context{}, nil, err
+		return nil, vm.Context{}, nil, fmt.Errorf("can not get the state of block %#x: %v", parent.Root(), err)
 	}
+	defer deferFn()
```
`deferFn()` is not always called.

ehnuje commented on node/cn/api_tracer.go:
```diff
@@ -769,6 +795,26 @@ func (api *PrivateDebugAPI) traceTx(ctx context.Context, message blockchain.Mess
+		logger.Debug("Get stateDB by computeStateDB", "block", block.NumberU64())
+		return stateDB, func() {}, nil
+	}
+	logger.Debug("Get stateDB from stateCache", "block", block.NumberU64())
```
How about including `reexec` to the log?

ehnuje commented on blockchain/blockchain.go:
```diff
@@ -1045,23 +1074,56 @@ func (bc *BlockChain) writeStateTrie(block *types.Block, state *state.StateDB) e
+				bc.triegc.Push(block.root, -float32(block.blockNum))
+				logger.Trace("Push GC block", "blkNum", block.blockNum, "hash", block.root.String())
+				blkNum := block.blockNum
+				if blkNum > triesInMemory {
```
```suggestion
				if blkNum <= triesInMemory {
				    continue
				}
```
This will reduce an indentation 😅

KimKyungup replied:
I updated. :)

kjhman21 commented on node/cn/api_tracer.go:
```diff
@@ -769,6 +795,26 @@ func (api *PrivateDebugAPI) traceTx(ctx context.Context, message blockchain.Mess
+	// If we have the state fully available, use that.
+	stateDB, err := api.cn.blockchain.StateAtWithGCLock(block.Root())
+	if err != nil {
+		// If no state is locally available, the desired state will be generated.
```
What about defining a function variable here?  e.g.,
```
var f = func() {}

...

return nil, f, err

...

return stateDB, f, nil
```

KimKyungup replied:
I changed it. :)

kjhman21 commented on storage/statedb/database.go:
```diff
@@ -541,6 +563,9 @@ func (db *Database) reference(child common.Hash, parent common.Hash) {
 
 // Dereference removes an existing reference from a root node.
 func (db *Database) Dereference(root common.Hash) {
+	db.gcLock.Lock()
```
hmm.. I'm not 100% sure, but it seems that `bd.lock` is for `db.nodes`.

Should we have a different lock for gc?

KimKyungup replied:
`database.lock` is for reading/writing only `database.nodes`.
But `gcLock` is for preventing to gc trie node in `database.nodes` without writing to persistent database.

In `StateAtWithGCLock`, with `RGCLock`, it returns the stateDB.
During the process, `DoesExistCachedNode` check if the state root is available with `database.lock`. 

After the function returns, the root can be removed by `Cap` or ` Commit`.
But in this cases, the root already was written on persistent database during `Cap`, `Commit`.

So, the stateDB is valid.

KimKyungup commented:
@kjhman21 @ehnuje @winnie-byun PTAL. :)

kjhman21 commented:
@KimKyungup Please add test cases and results for this PR.

KimKyungup commented:
I attached the test code in RPC console.
This code can try call traceBlock API repeatedly.

Before this PR, it always makes the issued situation.
But after this PR, the issued situation is resolved.

```
for (i=0;i<10;i++) {
	debug.traceBlockByNumber(klay.blockNumber,{tracer:"callTracer"})
	debug.traceBlockByNumber(klay.blockNumber-1,{tracer:"callTracer"})
	debug.traceBlockByNumber(klay.blockNumber-2,{tracer:"callTracer"})
	debug.traceBlockByNumber(klay.blockNumber-3,{tracer:"callTracer"})
	debug.traceBlockByNumber(klay.blockNumber-4,{tracer:"callTracer"})
}
```
